src/lib.rs

```diff
@@ -46,30 +47,80 @@ use pki_types::CertificateDer;
+///
+/// If **either** (or **both**) are set, certificates are only loaded from
+/// the locations specified via environment variables and not the platform-
+/// native certificate store.
+///
+/// ## Errors
+///
+/// This function fails in a platform-specific way, expressed in a [`std::io::Error`].
+///
+/// ## Caveats
 ///
 /// This function can be expensive: on some platforms it involves loading
 /// and parsing a ~300KB disk file.  It's therefore prudent to call
 /// this sparingly.
+///
+/// [c_rehash]: https://www.openssl.org/docs/manmaster/man1/c_rehash.html
 pub fn load_native_certs() -> Result<Vec<CertificateDer<'static>>, Error> {
-    load_certs_from_env().unwrap_or_else(platform::load_native_certs)
+    if let Some(certs) = load_certs_from_env()? {
+        return Ok(certs);
+    };
+    platform::load_native_certs()
 }
 
-const ENV_CERT_FILE: &str = "SSL_CERT_FILE";
-
-/// Returns None if SSL_CERT_FILE is not defined in the current environment.
+/// Returns certificates stored at SSL_CERT_FILE and/or SSL_CERT_DIR.
+///
+/// If neither is set, `None` is returned.
 ///
-/// If it is defined, it is always used, so it must be a path to an existing,
-/// accessible file from which certificates can be loaded successfully. However,
-/// [`rustls_pemfile::certs()`] will try to skip parts of a file which
-/// are not part of a certificate.
-fn load_certs_from_env() -> Option<Result<Vec<CertificateDer<'static>>, Error>> {
-    let cert_var_path = PathBuf::from(env::var_os(ENV_CERT_FILE)?);
+/// If SSL_CERT_FILE is defined, a file must exist at this path, and loading
+/// certificates from it must succeed. However, [`rustls_pemfile::certs()`]
+/// will try to skip parts of a file which are not part of a certificate, and the
+/// file may contain no certificates.
+///
+/// If SSL_CERT_DIR is defined, a directory must exist at this path, and all
+/// [hash files](`is_hash_file_name()`) contained in it must be loaded successfully,
+/// subject to the rules outlined above for SSL_CERT_FILE. The directory is not
+/// scanned recursively and may be empty.
+fn load_certs_from_env() -> Result<Option<Vec<CertificateDer<'static>>>, Error> {
+    let paths = CertPaths {
+        file: env::var_os(ENV_CERT_FILE).map(PathBuf::from),
+        dir: env::var_os(ENV_CERT_DIR).map(PathBuf::from),
+    };
+
+    if paths.file.is_none() && paths.dir.is_none() {
+        return Ok(None);
+    }
 
-    Some(load_pem_certs(&cert_var_path))
+    Ok(Some(load_certs_from_paths(&paths)?))
+}
+
+fn load_certs_from_paths(paths: &CertPaths) -> Result<Vec<CertificateDer<'static>>, Error> {
+    let mut certs = match &paths.file {
+        Some(cert_file) => load_pem_certs(cert_file)?,
+        None => Vec::new(),
+    };
+
+    if let Some(cert_dir) = &paths.dir {
+        certs.append(&mut load_pem_certs_from_dir(cert_dir)?);
+    }
+
+    certs.sort_unstable_by(|a, b| a.cmp(b));
+    certs.dedup();
+
+    Ok(certs)
+}
+
+struct CertPaths {
```

Comment:
I'm not sure this type is useful? We can just pass two arguments to `load_certs_from_paths()`?

Reply:
The reason I added the struct is that both arguments are of the same type. Calling `load_certs_from_paths(file, dir)` instead of `load_certs_from_paths(dir, file)` will compile and is hard to spot in reviews.

Want me to remove the type?

Reply:
Hmm, seems like a good rationale to keep it -- sorry for the noise!